packages/model/src/projection.ts

```diff
@@ -1,9 +1,321 @@
+ *          ProjectionKeys<Projection>
+ *          // -> never
+ *          ```
+ * @example ```ts
+ *          const model = object({ field1: number, field2: object({ inner1: string }) })
+ *          type Projection = InferProjection<typeof model>
+ *          ProjectionKeys<Projection>
+ *          // -> "field1" | "field2"
+ *          ```
+ */
+// prettier-ignore
+export type ProjectionKeys<P extends Projection>
+  = [P] extends [types.UnionType<{ all: types.LiteralType<true>, partial: types.ObjectType<'immutable', infer Ps extends Projections> }>] ? keyof Ps
+  : never
+
+/**
+ * Given a {@link Type type} `T` and one of the possible {@link ProjectionKeys keys of its projection},
+ * returns the subprojection corresponding to that key.
+ *
+ * @example ```ts
+ *          type Object = ObjectType<{ field1: NumberType, field2: NumberType }>
+ *          SubProjection<ObjectType, "field1">
+ *          // -> true
+ *          ```
+ */
+// prettier-ignore
+export type SubProjection<P extends Projection, K extends ProjectionKeys<P>>
+  = [P] extends [types.UnionType<{ all: types.LiteralType<true>, partial: types.ObjectType<'immutable', infer Ps extends Projections> }>] ? Ps[K]
+  : never
+
+/**
+ * @param projection the {@link Projection projection} to select a subprojection from
+ * @param key the key used to select a subprojection from the projection
+ * @returns the selected subprojection, if the provided key is `true` the result is always `true`
+ * @example ```ts
+ *          const model = object({ field: number })
+ *          subProjection(model, { field: true }, "field")
+ *          // -> true
+ *          ```
+ * @example ```ts
+ *          const model = object({ field: number })
+ *          subProjection(model, {}, "field")
+ *          // -> undefined
+ *          ```
+ * @example ```ts
+ *          const model = object({ field: number })
+ *          subProjection(model, {}, true)
+ *          // -> true
+ *          ```
+ */
+export function subProjection<const P extends Projection, K extends ProjectionKeys<P>>(
+  projection: P,
+  key: K,
+): SubProjection<P, K> {
+  return projection.kind === 'union'
+    ? (projection.variants.partial.types[key] as SubProjection<P, K>)
+    : failWithInternalError(
+        'It appears that `projection.subProjection` was called on a true projection with a key that should have been inferred as `never`',
+      )
+}
+
+/**
+ * @param projection the projection whose depth is returned
+ * @returns the depth of the projection, that is the maximum nesting of the projection
+ * @example ```ts
+ *          projectionDepth<typeof number>(true)
+ *          // -> 0
+ *          ```
+ * @example ```ts
+ *          const model = object({ field1: object({ inner1: number }), field2: number })
+ *          projectionDepth<typeof model>({ field1: { inner1: true } })
+ *          // -> 2
+ *          ```
+ */
+export function depth<P extends Projection>(projection: P): number {
```

Comment:
Needed but it should work with an instance of a projection and not a ProjectionType (copy from branch `develop`)